flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3340,6 +3341,99 @@ class ArrayExprLowering {
     return fir::ArrayBoxValue(tempRes, dest.getExtents());
   }
 
+  static ExtValue lowerLazyArrayExpression(
+      Fortran::lower::AbstractConverter &converter,
+      Fortran::lower::SymMap &symMap, Fortran::lower::StatementContext &stmtCtx,
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &expr,
+      mlir::Value var) {
+    ArrayExprLowering ael{converter, stmtCtx, symMap};
+    return ael.lowerLazyArrayExpression(expr, var);
+  }
+
+  ExtValue lowerLazyArrayExpression(
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &expr,
+      mlir::Value var) {
+    auto loc = getLoc();
+    // Once the loop extents have been computed, which may require being inside
+    // some explicit loops, lazily allocate the expression on the heap.
+    ccPrelude = [=](llvm::ArrayRef<mlir::Value> shape) -> mlir::Value {
+      auto load = builder.create<fir::LoadOp>(loc, var);
+      auto eleTy = fir::unwrapRefType(load.getType());
+      auto unknown = fir::SequenceType::getUnknownExtent();
+      fir::SequenceType::Shape extents(shape.size(), unknown);
+      auto seqTy = fir::SequenceType::get(extents, eleTy);
+      auto toTy = fir::HeapType::get(seqTy);
```

Comment:
This patterns seems to show up a few times below, it could maybe be moveed in a lambda/static helper: `getLazyArrayType(mlir::Type eleTy, unsigned rank)` ?

Reply:
I made an attempt at doing that, but then realized that although the pattern is the same, the different types and values are used/unused in different ways. I can look at it again though.